File: 新建文件夹/astralfox-rigging/core/pipeline.py
"""Full pipeline orchestration — separate → rig → export → deploy."""

import time
from loguru import logger

from core.separation import separate_layers
from core.rigging import rig_layers
from core.export import export_model
from core.deploy import deploy_model


def run_pipeline(
    image_id: str,
    template: str = "catgirl",
    mesh_density: str = "medium",
    auto_deploy: bool = False,
    target_name: str | None = None,
) -> dict:
    """Run the full AI rigging pipeline.

    Stages:
    1. Layer separation (SAM2/CPU)
    2. Skeleton rigging + mesh generation + weight assignment
    3. Cubism SDK export (moc3, model3.json, cmo3)
    4. (Optional) Deploy to target directory
    """
    start = time.perf_counter()

    # Stage 1: Separate
    logger.info(f"pipeline [{image_id}]: stage 1/4 — separating layers")
    sep_result = separate_layers(image_id)
    layers = sep_result["layers"]

    # Build layer dicts for rigging
    layer_dicts = [{"label": ly["label"], "bbox": ly["bbox"]} for ly in layers]

    # Stage 2: Rig
    logger.info(f"pipeline [{image_id}]: stage 2/4 — rigging ({len(layers)} layers)")
    rig_result = rig_layers(image_id, layer_dicts, template, mesh_density)

    # Stage 3: Export
    logger.info(f"pipeline [{image_id}]: stage 3/4 — exporting Cubism model")
    export_result = export_model(
        image_id=image_id,
        skeleton=rig_result["skeleton"],
        layers=layers,
        meshes=rig_result["meshes"],
        weights=rig_result["weights"],
    )

    # Stage 4: Deploy (optional)
    deploy_result = None
    if auto_deploy:
        logger.info(f"pipeline [{image_id}]: stage 4/4 — deploying")
        deploy_result = deploy_model(image_id, target_name=target_name)

    total_ms = int((time.perf_counter() - start) * 1000)
    logger.info(f"pipeline [{image_id}]: complete in {total_ms}ms")

    return {
        "separate": sep_result,
        "rig": rig_result,
        "export": export_result,
        "deploy": deploy_result,
        "total_time_ms": total_ms,
    }

File: 桌宠demo/新建文件夹/avatar-web-management/src/lib/rigging/pipeline.ts
// 管线编排器 — 分阶段调用 rigging 服务，实时推送进度
import { separateLayers, rigModel, exportModel, deployModel, downloadModelZip } from './client';
import { pushPipelineProgress } from '@/lib/ws/server';
import { createLogger } from '@/lib/logger';
import type { PipelineStage, PipelineOptions, PipelineStatus, LayerResult } from './types';

const log = createLogger('rigging-pipeline');

// In-memory status store (fallback when Redis unavailable)
const statusStore = new Map<string, PipelineStatus>();

export function getPipelineStatus(imageId: string): PipelineStatus | undefined {
  return statusStore.get(imageId);
}

function updateStatus(imageId: string, update: Partial<PipelineStatus>) {
  const existing = statusStore.get(imageId);
  const status: PipelineStatus = existing
    ? { ...existing, ...update }
    : {
        imageId,
        stage: 'uploading',
        percent: 0,
        message: 'Starting...',
        startedAt: Date.now(),
        ...update,
      };
  statusStore.set(imageId, status);

  // Push via WebSocket
  pushPipelineProgress(imageId, {
    stage: status.stage,
    percent: status.percent,
    message: status.message,
    error: status.error,
    result: status.result as Record<string, unknown> | undefined,
  });

  // Auto-cleanup after 1 hour
  setTimeout(() => statusStore.delete(imageId), 3_600_000);
}

export async function orchestratePipeline(
  imageId: string,
  options: PipelineOptions,
): Promise<PipelineStatus> {
  const startedAt = Date.now();
  updateStatus(imageId, { stage: 'uploading', percent: 0, message: '准备中...', startedAt });

  try {
    // Stage 1: Layer Separation (0-25%)
    updateStatus(imageId, { stage: 'separating', percent: 5, message: 'AI 图层分离中...' });
    log.info({ imageId }, 'Starting layer separation');

    const sepResult = await separateLayers(imageId);
    updateStatus(imageId, { stage: 'separating', percent: 25, message: `图层分离完成 (${sepResult.layers.length} 层, ${Math.round(sepResult.processingTimeMs / 1000)}s)` });

    // Stage 2: Rigging (25-55%)
    updateStatus(imageId, { stage: 'rigging', percent: 25, message: '骨骼预测中...' });
    log.info({ imageId, template: options.template }, 'Starting rigging');

    const layersForRig: { label: string; texture_url: string; mask_url: string; bbox: number[] }[] =
      sepResult.layers.map((l: LayerResult) => ({
        label: l.label,
        texture_url: l.textureUrl,
        mask_url: l.maskUrl,
        bbox: l.bbox,
      }));

    const rigResult = await rigModel(imageId, layersForRig, options.template, options.meshDensity);
    updateStatus(imageId, { stage: 'rigging', percent: 55, message: `骨骼绑定完成 (${rigResult.meshCount} 网格, ${Math.round(rigResult.processingTimeMs / 1000)}s)` });

    // Stage 3: Export (55-85%)
    updateStatus(imageId, { stage: 'exporting', percent: 55, message: 'Cubism 格式导出中...' });
    log.info({ imageId }, 'Starting export');

    const exportResult = await exportModel(
      imageId,
      rigResult.skeleton,
      layersForRig,
      rigResult.meshes,
      rigResult.weights,
    );
    updateStatus(imageId, { stage: 'exporting', percent: 85, message: `模型导出完成 (${Math.round(exportResult.processingTimeMs / 1000)}s)` });

    // Stage 4: Pull model files (85-95%)
    updateStatus(imageId, { stage: 'pulling_assets', percent: 85, message: '下载模型文件中...' });

    let modelZip: ArrayBuffer;
    try {
      modelZip = await downloadModelZip(imageId);
      updateStatus(imageId, { stage: 'pulling_assets', percent: 95, message: '模型文件下载完成' });
    } catch (err) {
      log.warn({ err, imageId }, 'Failed to download model ZIP — continuing without local copy');
      modelZip = new ArrayBuffer(0);
      updateStatus(imageId, { stage: 'pulling_assets', percent: 95, message: '模型文件下载跳过' });
    }

    // Stage 5: Deploy (95-100%) — optional
    let deployedPath: string | undefined;
    if (options.autoDeploy) {
      updateStatus(imageId, { stage: 'deploying', percent: 95, message: '部署到桌面端...' });
      try {
        const deployResult = await deployModel(imageId, undefined, options.targetName);
        deployedPath = deployResult.deployedPath;
        updateStatus(imageId, { stage: 'deploying', percent: 100, message: `部署完成: ${deployResult.deployedPath}` });
      } catch (err) {
        log.error({ err, imageId }, 'Deploy failed');
        updateStatus(imageId, { stage: 'deploying', percent: 100, message: '部署失败，模型已生成可手动部署' });
      }
    }

    const totalTimeMs = Date.now() - startedAt;
    const result = {
      modelId: imageId,
      previewUrl: exportResult.texturesUrls[0] || '',
      moc3Url: exportResult.moc3Url || exportResult.cmo3Url,
      totalTimeMs,
    };

    updateStatus(imageId, {
      stage: options.autoDeploy ? 'deploying' : 'pulling_assets',
      percent: 100,
      message: `完成! (${Math.round(totalTimeMs / 1000)}s)`,
      result: { ...result, deployedPath },
    });

    log.info({ imageId, totalTimeMs }, 'Pipeline completed successfully');
    return statusStore.get(imageId)!;

  } catch (err) {
    const errorMsg = (err as Error).message;
    log.error({ err, imageId }, 'Pipeline failed');
    updateStatus(imageId, { error: errorMsg, message: `失败: ${errorMsg}` });
    throw err;
  }
}
